src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs

```diff
@@ -104,15 +108,43 @@ public AuthorizeFilter(string policy)
 
         bool IFilterFactory.IsReusable => true;
 
-        /// <inheritdoc />
-        public virtual async Task OnAuthorizationAsync(AuthorizationFilterContext context)
+        private async Task<AuthorizationPolicy> GetEffectivePolicyAsync(AuthorizationFilterContext context)
         {
-            if (context == null)
-            {
-                throw new ArgumentNullException(nameof(context));
+            if (_effectivePolicy != null) {
+                return _effectivePolicy;
             }
 
             var effectivePolicy = Policy;
+
+            if (_mvcOptions == null) {
+                _mvcOptions = context.HttpContext.RequestServices.GetRequiredService<IOptions<MvcOptions>>().Value;
+            }
+
+            if (_mvcOptions.CombineAuthorizeFilters)
+            {
+                if (!context.IsEffectivePolicy<AuthorizeFilter>(this)) {
```

Comment:
This will cause issues as it will change the order in which the authorization filter executes. I'm going to give a practical example.

I plug in globally (in this order)

```csharp
new AuthorizeFilter(MyPolicy)
new AutoValidateAntiforgeryTokenFilter()
```

I add another Authorize filter in either my controller or my action.

```csharp
[Authorize("2FAPolicy")]
```
Now lets examine the existing and new behaviors:
With the existing behavior
The first authorize filter executes and evaluates the policy (Authenicating the user in the process)
The autovalidate antiforgery filter executes (with the identity provided by 1)
The second authorize filter executes (and sets up identity 2)

With the new behavior
The first authorize filter does nothing (returns).
The autovalidate antiforgery filter executes (with an anonymous identity)
The second authorize filter executes (and sets up a combined identity of 1 and 2)

What is wrong here is that the order in which the AutorizeFilter executes relative to other IAuthorizationFilter changes and that has an implicit impact on the execution of other filters. There are a couple of ways we can go about fixing this (we can do both as they complement each other and have value on their own):
* Instead of making the last filter do the combined authorization, we make the first filter combine the individual policies and execute the filter. (This is correct due to the fact that the semantics here are to combine instead of override).
* Make other IAuthorizationFilters in our platform ordered (Like ValidateAntiforgeryTokenFilter) that way this filter will always execute in the same order relative to those other two. That solves the problem for our built-in filters but not for custom implementations created by our users.

Reply:
Order takes precedence over scope.

This means that AF will run after **ALL** `AuthorizeFilter` instances

Reply:
@rynowak I didn't see https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.ViewFeatures/ValidateAntiForgeryTokenAttribute.cs#L21, so point 2 its covered then.

@rynowak Do we still care about doing
> What is wrong here is that the order in which the AutorizeFilter executes relative to other IAuthorizationFilter changes and that has an implicit impact on the execution of other filters.

This would only affect people building their own filters IAuthorizationFilter implementations that have a dependency on the authorize behavior. (and not ordering them). Do we care about it?

If we don't, then there's no point on making the first authorizefilter execute the combined policy instead of the last. On the other side, it should be cheap to change.

After we resolve this I think this is good to go, given that all our filters (CORS and Antiforgery) are ordered and won't run into any issue with this.

Reply:
> What is wrong here is that the order in which the AutorizeFilter executes relative to other IAuthorizationFilter changes and that has an implicit impact on the execution of other filters.

I would argue that it's right now, and used to be wrong. Anyone who cases about order of execution  relative to AuthorizeFilter needs to use `Order` - **End of story**. We've patched two issues related to that, which is too many 😆 

This will be an opt-in change, so it's not going to take anyone by surprise.

> If we don't, then there's no point on making the first authorizefilter execute the combined policy instead of the last. On the other side, it should be cheap to change.

The last filter should execute the policy. This is how filters work.